### PR TITLE
perf: Improve scope information collection performance

### PR DESCRIPTION
**Pull request by liuxingbaoyu**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

Here we introduced a simplified traversal, which improved the overall performance by about 10%. (Well, I don’t understand why it brings such a big performance improvement. I tried more later but failed. 🤦‍♂️)

```
PS F:\babel\benchmark> node --expose-gc .\all\real-case-ts-mjs.mjs 
all/real-case-ts-mjs.mjs babel-parser-express.ts @ current: 36.76 ops/sec ±8.46% 30 runs (27ms)
all/real-case-ts-mjs.mjs babel-parser-express.ts @ baseline: 33.86 ops/sec ±5.73% 30 runs (30ms)
all/real-case-ts-mjs.mjs ts-parser.ts @ current: 6.51 ops/sec ±6.65% 30 runs (154ms)
all/real-case-ts-mjs.mjs ts-parser.ts @ baseline: 5.21 ops/sec ±4.09% 30 runs (192ms)
PS F:\babel\benchmark> node --expose-gc .\all\real-case-mjs-cjs.mjs
all/real-case-mjs-cjs.mjs ts-checker.mjs @ current: 1.11 ops/sec ±2.37% 30 runs (903ms)
all/real-case-mjs-cjs.mjs ts-checker.mjs @ baseline: 0.93 ops/sec ±2.46% 30 runs (1079ms)
all/real-case-mjs-cjs.mjs ts-parser.mjs @ current: 8.29 ops/sec ±3.11% 30 runs (121ms)
all/real-case-mjs-cjs.mjs ts-parser.mjs @ baseline: 7.36 ops/sec ±3.63% 30 runs (136ms)
```

### CODE REVIEW

File: packages/babel-traverse/src/visitors.ts [hunk 1]
```diff
@@ -104,13 +104,9 @@ function explode$1<S>(visitor: Visitor<S>): ExplodedVisitor<S> {
     const types = virtualTypes[nodeType];
     if (types !== null) {
       for (const type of types) {
-        // merge the visitor if necessary or just put it back in
-        if (visitor[type]) {
-          mergePair(visitor[type], fns);
-        } else {
-          // @ts-expect-error Expression produces too complex union
-          visitor[type] = fns;
-        }
```
Comment (liuxingbaoyu) [inline packages/babel-traverse/src/visitors.ts:113]: This is a bug fix, previously it shallow copied `fns` to multiple places, but it just happened not to be modified. After we added the definition of `BlockScoped` it was modified later, causing the visitor to be wrong. Here we use `mergePair` to do a deep clone.

**Review by JLHwung — APPROVED** [review]
I think the performance improvement comes from the fact that we avoided creation of traversal context in the simplified traversal. We can investigate if we can share the traversal context used by the sub-traverseNode calls with the root context of the node path where we issued `path.traverse()` from. If it is possible, it will benefit all Babel plugins.

File: packages/babel-traverse/src/path/context.ts [hunk 2]
```diff
@@ -133,8 +133,8 @@ export function stop(this: NodePath) {
   this._traverseFlags |= SHOULD_SKIP | SHOULD_STOP;
 }
 
-export function setScope(this: NodePath) {
-  if (this.opts?.noScope) return;
+export function setScope(this: NodePath, ignoreNoScope?: boolean) {
```
Comment (nicolo-ribaudo) [inline packages/babel-traverse/src/path/context.ts:136]:
```suggestion
export function setScope(this: NodePath, forceScope?: boolean) {
```

`setScope` is (unfortunately) part of the public API. What do you think abuot adding an internal
```js
function _forceSetScope(path: NodePath) {}
```
instead?

Comment (liuxingbaoyu) [reply packages/babel-traverse/src/path/context.ts:136]: That's all good to me. :)

**Review by JLHwung — APPROVED** [review]

**Review by nicolo-ribaudo — APPROVED** [review]

**Comment by liuxingbaoyu** [comment]
@mydea Thanks for the report! I will take a look.

**Pull request merged**